File: app/src/main/java/de/gaffga/android/fragments/SessionEditFragment.java
package de.gaffga.android.fragments;

import android.app.Activity;
import androidx.fragment.app.Fragment;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import androidx.annotation.Nullable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import de.gaffga.android.zazentimer.DbOperations;
import de.gaffga.android.zazentimer.MessageView;
import de.gaffga.android.zazentimer.R;
import de.gaffga.android.zazentimer.ZazenTimerActivity;
import de.gaffga.android.zazentimer.bo.Section;
import de.gaffga.android.zazentimer.bo.Session;
import de.gaffga.android.zazentimer.databinding.FragmentEditSessionBinding;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.transition.MaterialSharedAxis;
import dagger.hilt.android.AndroidEntryPoint;
import java.util.Arrays;
import java.util.List;
import javax.inject.Inject;

@AndroidEntryPoint
public class SessionEditFragment extends Fragment {
    private static final String TAG = "ZMT_SessionEditFragment";
    private FragmentEditSessionBinding binding;
    private MessageView messageView;
    private SharedPreferences pref;
    private Section[] sections;
    private Session session = null;
    private int sessionId;
    private SectionListAdapter adapter;

    @Inject DbOperations dbOperations;

    private void handleAttach(Context context) {
    }

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        setEnterTransition(new MaterialSharedAxis(MaterialSharedAxis.X, true));
        setReturnTransition(new MaterialSharedAxis(MaterialSharedAxis.X, false));
        setHasOptionsMenu(true);
        if (bundle != null) {
            this.sessionId = bundle.getInt("sessionId");
        } else if (getArguments() != null) {
            this.sessionId = getArguments().getInt("sessionId");
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        if (menuItem.getItemId() == R.id.menu_session_edit_help) {
            showHelp13();
            return true;
        }
        return super.onOptionsItemSelected(menuItem);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        getActivity().getMenuInflater().inflate(R.menu.session_edit_menu, menu);
        super.onPrepareOptionsMenu(menu);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        handleAttach(context);
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        handleAttach(activity);
    }

    @Override
    @Nullable
    public View onCreateView(LayoutInflater layoutInflater, ViewGroup viewGroup, Bundle bundle) {
        Log.d(TAG, "onCreateView");
        binding = FragmentEditSessionBinding.inflate(layoutInflater, viewGroup, false);
        this.pref = ZazenTimerActivity.getPreferences(getActivity());

        adapter = new SectionListAdapter(new SectionListAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(Section section) {
                SessionEditFragment.this.navigateToSectionEdit(section.id);
            }
        });

        binding.list.setLayoutManager(new LinearLayoutManager(getActivity()));
        binding.list.setAdapter(adapter);

        SectionTouchHelperCallback callback = new SectionTouchHelperCallback(
            new SectionTouchHelperCallback.SectionTouchListener() {
                @Override
                public void onSwipe(int position) {
                    final Section deletedSection = adapter.getItem(position);
                    final int deletedPosition = position;
                    dbOperations.deleteSection(deletedSection.id);
                    adapter.removeItem(position);

                    Snackbar.make(binding.list, "Deleted '" + deletedSection.toString() + "'", Snackbar.LENGTH_LONG)
                        .setAction("UNDO", new View.OnClickListener() {
                            @Override
                            public void onClick(View view) {
                                dbOperations.insertSection(SessionEditFragment.this.session, deletedSection);
                                adapter.insertItem(deletedPosition, deletedSection);
                            }
                        })
                        .show();
                }

                @Override
                public boolean onMove(int fromPosition, int toPosition) {
                    adapter.moveItem(fromPosition, toPosition);
                    return true;
                }
            }
        );
        new ItemTouchHelper(callback).attachToRecyclerView(binding.list);

        binding.butNewSection.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                SessionEditFragment.this.doCreateNewSection();
            }
        });
        return binding.getRoot();
    }

    @Override
    public void onSaveInstanceState(Bundle bundle) {
        bundle.putInt("sessionId", this.sessionId);
    }

    public void setSessionId(int i) {
        this.sessionId = i;
    }

    private void initSectionList() {
        adapter.setItems(Arrays.asList(this.sections));
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.d(TAG, "sessionId=" + this.sessionId);
        this.session = dbOperations.readSession(this.sessionId);
        if (this.session == null) {
            Log.e(TAG, "session is NULL");
        } else {
            Log.i(TAG, "session found and valid");
        }
        binding.textSitzungName.setText(this.session.name);
        binding.textSitzungBeschreibung.setText(this.session.description);
        getActivity().invalidateOptionsMenu();
        this.sections = dbOperations.readSections(this.session.id);
        initSectionList();
        binding.textSitzungName.setText(this.session.name);
        binding.textSitzungBeschreibung.setText(this.session.description);
        if (this.pref.getBoolean(ZazenTimerActivity.PREF_KEY_SHOW_SESSION_EDIT_HELP_V13, false)) {
            showHelp13();
            this.pref.edit().putBoolean(ZazenTimerActivity.PREF_KEY_SHOW_SESSION_EDIT_HELP_V13, true).apply();
        }
    }

    private void showHelp13() {
        if (this.messageView != null) {
            return;
        }
        this.messageView = new MessageView(getActivity());
        this.messageView.setTitle(getString(R.string.help_sectionlist_title));
        this.messageView.setText(getString(R.string.help_sectionlist_text));
        this.messageView.setOnOkListener(new Runnable() {
            @Override
            public void run() {
                SessionEditFragment.this.messageView = null;
            }
        });
        this.messageView.show();
    }

    @Override
    public void onPause() {
        super.onPause();
        List<Section> items = adapter.getItems();
        for (int i = 0; i < items.size(); i++) {
            Section section = items.get(i);
            section.rank = i + 1;
            dbOperations.updateSection(section);
        }
        this.session.name = binding.textSitzungName.getText().toString();
        this.session.description = binding.textSitzungBeschreibung.getText().toString();
        dbOperations.updateSession(this.session);
    }

    public void doCreateNewSection() {
        Section section = new Section(getResources().getString(R.string.default_section_name), 60);
        dbOperations.insertSection(this.session, section);
        navigateToSectionEdit(section.id);
    }

    private void navigateToSectionEdit(int sectionId) {
        Bundle args = new Bundle();
        args.putInt("sectionId", sectionId);
        Navigation.findNavController(getView()).navigate(R.id.action_sessionEditFragment_to_sectionEditFragment, args);
    }
}

File: app/src/main/kotlin/at/priv/graf/zazentimer/fragments/SessionEditFragment.kt
package at.priv.graf.zazentimer.fragments

import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.navigation.Navigation
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import at.priv.graf.zazentimer.R
import at.priv.graf.zazentimer.ZazenTimerActivity
import at.priv.graf.zazentimer.bo.Section
import at.priv.graf.zazentimer.bo.Session
import at.priv.graf.zazentimer.database.DbOperations
import at.priv.graf.zazentimer.databinding.FragmentEditSessionBinding
import at.priv.graf.zazentimer.views.MessageView
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.transition.MaterialSharedAxis
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import javax.inject.Inject

@AndroidEntryPoint
class SessionEditFragment : Fragment() {
    private var _binding: FragmentEditSessionBinding? = null
    private val binding get() = _binding!!

    private var messageView: MessageView? = null
    private var pref: SharedPreferences? = null
    private var sections: Array<Section>? = null
    private var session: Session? = null
    private var sessionId: Int = 0
    private var adapter: SectionListAdapter? = null

    @Inject
    lateinit var dbOperations: DbOperations

    override fun onCreate(bundle: Bundle?) {
        super.onCreate(bundle)
        enterTransition = MaterialSharedAxis(MaterialSharedAxis.X, true)
        returnTransition = MaterialSharedAxis(MaterialSharedAxis.X, false)
        if (bundle != null) {
            this.sessionId = bundle.getInt("sessionId")
        } else if (arguments != null) {
            this.sessionId = requireArguments().getInt("sessionId")
        }
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        super.onViewCreated(view, savedInstanceState)
        requireActivity().addMenuProvider(
            object : androidx.core.view.MenuProvider {
                override fun onCreateMenu(
                    menu: Menu,
                    menuInflater: MenuInflater,
                ) {
                    menuInflater.inflate(R.menu.session_edit_menu, menu)
                }

                override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                    if (menuItem.itemId == R.id.menu_session_edit_help) {
                        editHelper.showHelp13()
                        return true
                    }
                    return false
                }
            },
            viewLifecycleOwner,
            androidx.lifecycle.Lifecycle.State.RESUMED,
        )
    }

    override fun onCreateView(
        layoutInflater: LayoutInflater,
        viewGroup: ViewGroup?,
        bundle: Bundle?,
    ): View {
        Log.d(TAG, "onCreateView")
        _binding = FragmentEditSessionBinding.inflate(layoutInflater, viewGroup, false)
        this.pref = ZazenTimerActivity.getPreferences(requireContext())

        adapter =
            SectionListAdapter(
                object : SectionListAdapter.OnItemClickListener {
                    override fun onItemClick(section: Section) {
                        this@SessionEditFragment.editHelper.navigateToSectionEdit(section.id)
                    }
                },
                object : SectionListAdapter.OnSectionActionListener {
                    override fun onDeleteSection(position: Int) {
                        deleteSectionAt(position)
                    }

                    override fun onDuplicateSection(position: Int) {
                        duplicateSectionAt(position)
                    }
                },
            )

        binding.list.layoutManager = LinearLayoutManager(requireContext())
        binding.list.adapter = adapter

        val callback =
            SectionTouchHelperCallback(
                object : SectionTouchHelperCallback.SectionTouchListener {
                    override fun onSwipe(position: Int) {
                        deleteSectionAt(position)
                    }

                    override fun onMove(
                        fromPosition: Int,
                        toPosition: Int,
                    ): Boolean {
                        adapter?.moveItem(fromPosition, toPosition) ?: return false
                        return true
                    }
                },
            )
        ItemTouchHelper(callback).attachToRecyclerView(binding.list)

        binding.butNewSection.setOnClickListener {
            editHelper.doCreateNewSection()
        }
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onSaveInstanceState(bundle: Bundle) {
        bundle.putInt("sessionId", this.sessionId)
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "sessionId=${this.sessionId}")
        lifecycleScope.launch {
            this@SessionEditFragment.session = dbOperations.readSession(this@SessionEditFragment.sessionId)
            if (!isAdded) return@launch
            val s = this@SessionEditFragment.session
            if (s == null) {
                Log.e(TAG, "session is NULL")
            } else {
                Log.i(TAG, "session found and valid")
                binding.textSitzungName.setText(s.name)
                binding.textSitzungBeschreibung.setText(s.description)
                this@SessionEditFragment.sections = dbOperations.readSections(s.id)
                editHelper.initSectionList()
            }
        }
        requireActivity().invalidateOptionsMenu()
        if (this.pref?.getBoolean(ZazenTimerActivity.PREF_KEY_SHOW_SESSION_EDIT_HELP_V13, false) == false) {
            editHelper.showHelp13()
            this.pref
                ?.edit()
                ?.putBoolean(ZazenTimerActivity.PREF_KEY_SHOW_SESSION_EDIT_HELP_V13, true)
                ?.apply()
        }
    }

    override fun onPause() {
        super.onPause()
        val items = adapter?.getItems() ?: return
        val sectionsToUpdate = ArrayList<Section>()
        for (i in items.indices) {
            val section = items[i]
            section.rank = i + 1
            sectionsToUpdate.add(section)
        }
        session?.let { s ->
            s.name = binding.textSitzungName.text.toString()
            s.description = binding.textSitzungBeschreibung.text.toString()
            runBlocking {
                for (section in sectionsToUpdate) {
                    dbOperations.updateSection(section)
                }
                dbOperations.updateSession(s)
            }
        }
    }

    private fun deleteSectionAt(position: Int) {
        val a = adapter ?: return
        val s = session ?: return
        val deletedSection = a.getItem(position)
        val deletedPosition = position
        lifecycleScope.launch {
            dbOperations.deleteSection(deletedSection.id.toLong())
        }
        a.removeItem(position)

        Snackbar
            .make(binding.list, getString(R.string.section_deleted, deletedSection), Snackbar.LENGTH_LONG)
            .setAction(getString(R.string.action_undo)) {
                lifecycleScope.launch {
                    dbOperations.insertSection(s, deletedSection)
                    a.insertItem(deletedPosition, deletedSection)
                }
            }.show()
    }

    private fun duplicateSectionAt(position: Int) {
        val a = adapter ?: return
        val s = session ?: return
        val source = a.getItem(position)
        val copy = Section(source.name ?: "", source.duration)
        copy.bell = source.bell
        copy.bellUri = source.bellUri
        copy.bellcount = source.bellcount
        copy.bellpause = source.bellpause
        copy.volume = source.volume
        lifecycleScope.launch {
            dbOperations.insertSection(s, copy)
            sections = dbOperations.readSections(s.id)
            editHelper.initSectionList()
        }
    }

    private val editHelper by lazy { SessionEditNavigationHelper() }

    private inner class SessionEditNavigationHelper {
        fun initSectionList() {
            adapter?.setItems(sections?.toList() ?: emptyList())
        }

        fun showHelp13() {
            if (messageView != null) {
                return
            }
            messageView = MessageView(requireActivity())
            messageView?.let { mv ->
                mv.setTitle(getString(R.string.help_sectionlist_title))
                mv.setText(getString(R.string.help_sectionlist_text))
                mv.setOnOkListener(
                    Runnable {
                        messageView = null
                    },
                )
                mv.show()
            }
        }

        fun doCreateNewSection() {
            val s = session ?: return
            val section =
                Section(
                    resources.getString(R.string.default_section_name),
                    DEFAULT_SECTION_DURATION_SECONDS,
                )
            lifecycleScope.launch {
                dbOperations.insertSection(s, section)
                navigateToSectionEdit(section.id)
            }
        }

        fun navigateToSectionEdit(sectionId: Int) {
            val args = Bundle()
            args.putInt("sectionId", sectionId)
            Navigation
                .findNavController(requireView())
                .navigate(R.id.action_sessionEditFragment_to_sectionEditFragment, args)
        }
    }

    companion object {
        private const val TAG = "ZMT_SessionEditFragment"
        private const val DEFAULT_SECTION_DURATION_SECONDS = 60
    }
}
